Source/AwsCommonRuntimeKit/crt/CBOR.swift

```diff
@@ -0,0 +1,360 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+import AwsCCommon
+import Foundation
+
+/// CBOR Types. These types don't map one-to-one to the CBOR RFC.
+public enum CBORType: Equatable {
+    /// UINT64 type for positive numbers.
+    case uint64(_ value: UInt64)
+    /// INT64 type for negative numbers. If the number is positive, it will be encoded as UINT64 type.
+    case int(_ value: Int64)
+    /// Double type. It might be encoded as an integer if possible without loss of precision. Half-precision floats are not supported.
+    case double(_ value: Double)
```

Comment:
worth to mention that the numbers will be encoded as "smallest possible".

Reply:
I don't get why we have `uint64` and `int`

Either we just do `int64` for both positive or negative, which means we only supports [-2^63; 2^63-1] range. (If SDKs are okay with that.)

Or we be clear about `unsgined 64` and `negative 64`, which both take unsigned int 64 to meet the range cbor supports, you can see [here](https://www.rfc-editor.org/rfc/rfc8949.html#name-major-types), cbor supports range from  [-2^64; 2^64-1] .

Or we do both, be clear about unsigned and negative, and then a helper to just take int. So that we can support the range cbor required, while provides an easier to use API.

Reply:
Thanks, I didn't realize `-2^64` is a possibility. I think we should just keep it simple and expose Int64 and UInt64 since they are Swift types. Maybe we can use better names like `uint` and `int`?
 
> Either we just do int64 for both positive or negative, which means we only supports [-2^63; 2^63-1] range. (If SDKs are okay with that.)

UInt64.max is a valid value, so we should support encoding/decoding that. 

> Or we be clear about unsgined 64 and negative 64, which both take unsigned int 64 to meet the range cbor supports, you can see [here](https://www.rfc-editor.org/rfc/rfc8949.html#name-major-types), cbor supports range from [-2^64; 2^64-1] .

The problem with `-2^64` is that there is no way to represent that in Swift in a type-safe way. During encoding, I don't think there is any way for Swift to have a `-2^64` integer unless it does what CBOR is doing which is `UInt64 named NegInt`. 

During decoding, yeah, we might encounter `-2^64` since CBOR allows it, but we should just error out instead of passing that complexity to the users to deal with. This should never happen in practice because I don't think anyone would be using `-2^64` numbers. We can add the `UInt64 named NegInt` type in the future if needed. 

@dayaffe What do you think?

Reply:
Actually I thought of that again.

It seems to be a one-way door as it's an enum, that we cannot add the `UInt64 named NegInt` type in the future, since how will you deal with the value that's possible to fit into two different types??

Reply:
Like currently, looks like you have both int64 and uint64 for the positive number, but during decode, you only give the negative value to the int64, and all the positive value will be uin64. 

It's arguably working as expected? But, seems more reasonable to just have two different type for negative and positive?

Maybe something like
```
public func fromInt64(int) throws -> CBORType {
}
``` 
???

Reply:
Discussed offline, it's not a one-way door since we can add a negint type in the future. I have added a warning that this enum is non_exhaustive. It is not ideal because we can encode it as different types, but we already have this issue with bytes, text, and double types, etc., so it's not too bad.

> as a helper to take the int and convert it to the type???

Yeah, we can do that, but I am not too concerned about the encoding part. The problem is during decoding; we can't map -2^64 to a Swift type, and I think we should try to hide this complexity from our users until needed. The API I am trying to design is to just encode Swift types and decode to Swift types to keep it simple and expected.